src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -185,15 +185,30 @@ public static unsafe void MatMulX(float[] mat, float[] src, float[] dst, int cro
                             Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
 
                             float* pMatTemp = pMatCurrent;
-                            Vector256<float> x01 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp));
-                            Vector256<float> x11 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> x21 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> x31 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp += ccol));
+                            if (Fma.IsSupported)
```

Comment:
It might be nice to abstract this into a "MultiplyAdd" helper, which itself contains the `Fma.IsSupported` check and either calls `Fma.MultiplyAdd` or does the individual `Multiply` and `Add` operations.

That would make this outer code cleaner, as we could just have four calls to the helper method (and hopefully the JIT would do the right thing for codegen here).

Reply:
Good idea. I also wonder whether the repeated checks of `Fma.IsSupported` create efficient code. Just looking at the code, a single check per invocation seems possible, or, even better, one per process instantiation. Then again, the JIT might do that for us?

Reply:
The JIT treats the `IsSupported` checks as constant and will completely drop the irrelevant path from the generated code.

Reply:
My only concern with creating a helper method here is that the JIT will enforce the evaluation order and will no longer fold the respective `LoadVector256` into the `Fma.MultiplyAdd` or `Avx.Multiply` calls....

There are probable workarounds if that is the case (such as passing the pointer we want to load from), but it is worth checking (and possibly logging a bug if it doesn't "just work").

Reply:
> The JIT treats the IsSupported checks as constant and will completely drop the irrelevant path from the generated code.

Thanks for explaining! This is cool :)

Reply:
> It might be nice to abstract this into a "MultiplyAdd" helper, which itself contains the `Fma.IsSupported` check and either calls `Fma.MultiplyAdd` or does the individual `Multiply` and `Add` operations.
> 
> That would make this outer code cleaner, as we could just have four calls to the helper method (and hopefully the JIT would do the right thing for codegen here).

Your concern are right. The sample code `MyMulAdd` does not fold the load into FMA. We need to use something like `MyUnsafeMulAdd`. I will change the code accordingly. 

```
        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
        private static Vector256<float> MyMulAdd(Vector256<float> a, Vector256<float> b, Vector256<float> c)
        {
            return Fma.MultiplyAdd(a, b, c);
        }
```

```
        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
        private unsafe static Vector256<float> MyUnsafeMulAdd(float* ptra, Vector256<float> b, Vector256<float> c)
        {
            return Fma.MultiplyAdd(Avx.LoadVector256(ptra), b, c);
        }
```